notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,526 @@
+Axon.Display.as_table(model, Nx.template({4, 1, 28, 28}, :f32)) |> IO.puts()
+```
+
+```elixir
+batch_size = 128
+
+train_data = Data.prepare_training_data(train_images, 128)
+test_data = Data.prepare_training_data(test_images, 128)
+
+{input_batch, target_batch} = Enum.at(train_data, 0)
+Kino.render(input_batch[[images: 0]] |> Data.image_to_kino())
+Kino.render(target_batch[[images: 0]] |> Data.image_to_kino())
+
+:ok
+```
+
+When training, it can be useful to stop execution early - either when you see it's failing and you don't want to waste time waiting for the remaining epochs to finish, or if it's good enough and you want to start experimenting with it.
+
+The `kino_early_stop/1` function below is a handy handler to give us a `Kino.Control.button` that will stop the training loop when clicked.
+
+We also have `plot_losses/1` function to visualize our train and validation losses using `VegaLite`.
+
+```elixir
+defmodule KinoAxon do
+  @doc """
+  Adds handler function which adds a frame with a "stop" button 
+  to the cell with the training loop.
+
+  Clicking "stop" will halt the training loop.
+  """
+  def kino_early_stop(loop) do
+    frame = Kino.Frame.new() |> Kino.render()
+    stop_button = Kino.Control.button("stop")
+    Kino.Frame.render(frame, stop_button)
+
+    {:ok, button_agent} = Agent.start_link(fn -> nil end)
+
+    Kino.Control.stream(stop_button, fn _event ->
+      Agent.update(button_agent, fn _ -> :stop end)
+    end)
+
+    handler = fn state ->
+      stop_state = Agent.get(button_agent, & &1)
+
+      if stop_state == :stop do
+        Agent.stop(button_agent)
+        Kino.Frame.render(frame, "stopped")
+        {:halt_loop, state}
+      else
+        {:continue, state}
+      end
+    end
+
+    Axon.Loop.handle(loop, :iteration_completed, handler)
+  end
+
+  @doc """
+  Plots the training and validation losses using Kino and VegaLite.
+
+  This *must* come after `Axon.Loop.validate`.
+  """
+  def plot_losses(loop) do
+    vl_widget =
+      Vl.new(width: 600, height: 400)
+      |> Vl.mark(:point, tooltip: true)
+      |> Vl.encode_field(:x, "epoch", type: :ordinal)
+      |> Vl.encode_field(:y, "loss", type: :quantitative)
+      |> Vl.encode_field(:color, "dataset", type: :nominal)
+      |> Kino.VegaLite.new()
+      |> Kino.render()
+
+    # |> Vl.layers([
+    #   Vl.new()
+    #   |> Vl.mark(:point, tooltip: true)
+    #   |> Vl.encode_field(:x, "epoch", type: :ordinal)
+    #   |> Vl.encode_field(:y, "loss", type: :quantitative),
+    #   Vl.new()
+    #   |> Vl.mark(:point, tooltip: true)
+    #   |> Vl.encode_field(:x, "epoch", type: :ordinal)
+    #   |> Vl.encode_field(:y, "val_loss", type: :quantitative)
+    # ])
+    # |> Kino.VegaLite.new()
+    # |> Kino.render()
+
+    handler = fn state ->
+      %Axon.Loop.State{metrics: metrics, epoch: epoch} = state
+      loss = metrics["loss"] |> Nx.to_number()
+      val_loss = metrics["validation_0"]["loss"] |> Nx.to_number()
```

Comment:
Looking at this we should probably make the validation metrics show up as a nested map like:

```
%{"validation" => %{0 => %{."loss" => ...}}}
```

Do you think that would be more intuitive to work with?

Reply:
I think my main confusion was what the `0` represented. When I first wrote it I assumed it represented the epoch. Either one seems fine, your alternative seems a little cleaner